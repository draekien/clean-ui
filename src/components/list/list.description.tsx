// eslint-disable-next-line tsdoc/syntax
/** @jsxImportSource theme-ui */
import React, { Fragment } from 'react';
import { DirectionProps, ListData } from './list';
import { descriptionTitleCss, listContainerCss, listItemCss } from './list.styles';

export interface DescriptionListProps extends DirectionProps {
  data: ListData[];
}
export const DescriptionList: React.FC<DescriptionListProps> = ({
  data,
  direction = 'column',
}) => {
  return (
    <dl sx={listContainerCss}>
      {data.map(({ label, rows }, i) => {
        return (
          <Fragment key={i}>
            <dt sx={descriptionTitleCss}>{label}</dt>
            {rows.map((row, j) => (
              <dd key={`${i}-${j}`} sx={listItemCss(direction)}>
                {row}
              </dd>
            ))}
          </Fragment>
        );
      })}
    </dl>
  );
};
