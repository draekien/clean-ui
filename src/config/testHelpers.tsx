import React, { ReactNode } from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from '../components/theme/theme.provider';

export const themedRender = (element: ReactNode) =>
  render(<ThemeProvider>{element}</ThemeProvider>);
