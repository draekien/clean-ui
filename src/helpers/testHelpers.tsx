import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from '../components/theme/theme.provider';

export const themedRender = (element: React.ReactNode) => {
  return render(<ThemeProvider>{element}</ThemeProvider>);
};

export const hexToRgb = (hex: string) => {
  hex = hex.replace(/[^0-9A-F]/gi, '');
  const bigint = parseInt(hex, 16);
  const r = (bigint >> 16) & 255;
  const g = (bigint >> 8) & 255;
  const b = bigint & 255;

  return `rgb(${r}, ${g}, ${b})`;
};
