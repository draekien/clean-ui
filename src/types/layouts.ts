export type HorizontalAlignment = 'left' | 'center' | 'right';
export type VerticalAlignment = 'top' | 'center' | 'bottom';
export type Direction = 'row' | 'column';
export type Size = 'small' | 'medium' | 'large';
