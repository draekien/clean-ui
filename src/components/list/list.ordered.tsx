/** @jsxImportSource theme-ui */
import React from 'react';
import { DirectionProps, ListData } from './list';
import {
  listContainerCss,
  orderedListItemContentCss,
  orderedListItemCss,
  orderedListItemIndexCss,
} from './list.styles';

export interface OrderedListProps extends DirectionProps {
  data: ListData;
}

export const OrderedList: React.FC<OrderedListProps> = ({
  data,
  direction = 'column',
}) => {
  return (
    <ol sx={listContainerCss}>
      {data.rows.map((row, index) => (
        <li key={index} sx={orderedListItemCss(direction)}>
          <div sx={orderedListItemIndexCss}>{`${index + 1}.`}</div>
          <div sx={orderedListItemContentCss}>{row}</div>
        </li>
      ))}
    </ol>
  );
};
