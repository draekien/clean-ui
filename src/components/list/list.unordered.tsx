// eslint-disable-next-line tsdoc/syntax
/** @jsxImportSource theme-ui */
import React from 'react';
import { DirectionProps, ListData } from './list';
import { listContainerCss, listItemCss } from './list.styles';

export interface UnorderedListProps extends DirectionProps {
  data: ListData;
}

export const UnorderedList: React.FC<UnorderedListProps> = ({
  data,
  direction = 'column',
}) => {
  return (
    <ul sx={listContainerCss}>
      {data.rows.map((row, index) => (
        <li key={index} sx={listItemCss(direction)}>
          {row}
        </li>
      ))}
    </ul>
  );
};
