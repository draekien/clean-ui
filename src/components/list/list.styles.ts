import { ThemeUIStyleObject } from '@theme-ui/css';
import { Direction } from '../../types/layouts';

export const listContainerCss: ThemeUIStyleObject = {
  padding: 0,
  margin: 0,
};

export const descriptionTitleCss: ThemeUIStyleObject = {
  variant: 'text.subtitle',
  paddingBottom: 'sm',
};

export const listItemCss = (direction: Direction): ThemeUIStyleObject => {
  const css: ThemeUIStyleObject = {
    padding: 0,
    margin: 0,
    display: 'inline-flex',
    variant: 'text.body',
  };

  if (direction === 'column') {
    css.display = 'flex';
  }

  return css;
};

export const orderedListItemCss = (direction: Direction): ThemeUIStyleObject => {
  const css: ThemeUIStyleObject = {
    padding: 0,
    margin: 0,
    display: 'inline-flex',
    variant: 'text.body',
  };

  if (direction === 'column') {
    css.display = 'flex';
  }

  return css;
};

export const orderedListItemIndexCss: ThemeUIStyleObject = {
  width: '4em',
  variant: 'text.mono',
  textAlign: 'center',
};

export const orderedListItemContentCss: ThemeUIStyleObject = {
  flexGrow: 1,
};
