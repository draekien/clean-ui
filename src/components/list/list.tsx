/** @jsxImportSource theme-ui */
import React from 'react';
import { Direction } from '../../types/layouts';
import { DescriptionList } from './list.description';
import { OrderedList } from './list.ordered';
import { UnorderedList } from './list.unordered';

export type ListVariant = 'ordered' | 'unordered' | 'description';
export interface ListData {
  label?: React.ReactNode;
  rows: React.ReactNode[];
}

export interface DirectionProps {
  direction?: Direction;
}

export interface ListProps extends DirectionProps {
  /** the type of list to render
   *  @defaultValue 'unordered'
   */
  variant?: ListVariant;
  data: ListData | ListData[];
}

export const List: React.FC<ListProps> = ({
  variant = 'unordered',
  direction = 'column',
  data,
}) => {
  switch (variant) {
    case 'description':
      return (
        <DescriptionList
          data={Array.isArray(data) ? data : [data]}
          direction={direction}
        />
      );
    case 'ordered':
      return (
        <OrderedList data={Array.isArray(data) ? data[0] : data} direction={direction} />
      );
    default:
      return (
        <UnorderedList
          data={Array.isArray(data) ? data[0] : data}
          direction={direction}
        />
      );
  }
};

export * from './list.description';
export * from './list.ordered';
export * from './list.unordered';
