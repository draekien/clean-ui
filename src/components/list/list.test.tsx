import React from 'react';
import { List, ListData, ListVariant } from './list';
import { themedRender } from '../../helpers/testHelpers';
import { Direction } from '../../types/layouts';

describe('List Component', () => {
  const variants: ListVariant[] = ['description', 'ordered', 'unordered'];
  const directions: Direction[] = ['column', 'row'];
  const data: ListData = {
    label: 'Critical Role Cast Members',
    rows: [
      'Travis Willingham',
      'Laura Bailey',
      'Ashleigh Johnson',
      'Sam Riegal',
      'Matthew Mercer',
      'Marisha Ray',
      "Liam O'Brian",
      'Talisen Jaffe',
    ],
  };

  variants.forEach((variant) => {
    it(`${variant} should match snapshot`, () => {
      const { container } = themedRender(<List variant={variant} data={data} />);

      expect(container).toMatchSnapshot();
    });
  });

  directions.forEach((direction) => {
    variants.forEach((variant) => {
      it(`${variant} renders correctly for the ${direction} direction`, () => {
        const { container } = themedRender(
          <List variant={variant} data={data} direction={direction} />
        );

        let query: string;

        switch (variant) {
          case 'description':
            query = 'dd';
            break;
          default:
            query = 'li';
            break;
        }

        const styles = window.getComputedStyle(container.querySelector(query) as Element);
        if (direction === 'column') {
          expect(styles.display).toBe('flex');
        } else {
          expect(styles.display).toBe('inline-flex');
        }
      });
    });
  });
});
