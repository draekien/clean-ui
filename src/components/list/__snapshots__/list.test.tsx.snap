// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List Component description should match snapshot 1`] = `
<div>
  <dl
    class="css-2mry40-DescriptionList"
  >
    <dt
      class="css-1fgqwsw"
    >
      Critical Role Cast Members
    </dt>
    <dd
      class="css-131zfi8"
    >
      Travis Willingham
    </dd>
    <dd
      class="css-131zfi8"
    >
      Laura Bailey
    </dd>
    <dd
      class="css-131zfi8"
    >
      Ashleigh Johnson
    </dd>
    <dd
      class="css-131zfi8"
    >
      Sam Riegal
    </dd>
    <dd
      class="css-131zfi8"
    >
      Matthew Mercer
    </dd>
    <dd
      class="css-131zfi8"
    >
      Marisha Ray
    </dd>
    <dd
      class="css-131zfi8"
    >
      Liam O'Brian
    </dd>
    <dd
      class="css-131zfi8"
    >
      Talisen Jaffe
    </dd>
  </dl>
</div>
`;

exports[`List Component ordered should match snapshot 1`] = `
<div>
  <ol
    class="css-1mhgfof-OrderedList"
  >
    <li
      class="css-131zfi8"
    >
      <div
        class="css-1wa42mg"
      >
        1.
      </div>
      <div
        class="css-6wajs7"
      >
        Travis Willingham
      </div>
    </li>
    <li
      class="css-131zfi8"
    >
      <div
        class="css-1wa42mg"
      >
        2.
      </div>
      <div
        class="css-6wajs7"
      >
        Laura Bailey
      </div>
    </li>
    <li
      class="css-131zfi8"
    >
      <div
        class="css-1wa42mg"
      >
        3.
      </div>
      <div
        class="css-6wajs7"
      >
        Ashleigh Johnson
      </div>
    </li>
    <li
      class="css-131zfi8"
    >
      <div
        class="css-1wa42mg"
      >
        4.
      </div>
      <div
        class="css-6wajs7"
      >
        Sam Riegal
      </div>
    </li>
    <li
      class="css-131zfi8"
    >
      <div
        class="css-1wa42mg"
      >
        5.
      </div>
      <div
        class="css-6wajs7"
      >
        Matthew Mercer
      </div>
    </li>
    <li
      class="css-131zfi8"
    >
      <div
        class="css-1wa42mg"
      >
        6.
      </div>
      <div
        class="css-6wajs7"
      >
        Marisha Ray
      </div>
    </li>
    <li
      class="css-131zfi8"
    >
      <div
        class="css-1wa42mg"
      >
        7.
      </div>
      <div
        class="css-6wajs7"
      >
        Liam O'Brian
      </div>
    </li>
    <li
      class="css-131zfi8"
    >
      <div
        class="css-1wa42mg"
      >
        8.
      </div>
      <div
        class="css-6wajs7"
      >
        Talisen Jaffe
      </div>
    </li>
  </ol>
</div>
`;

exports[`List Component unordered should match snapshot 1`] = `
<div>
  <ul
    class="css-pg1eyq-UnorderedList"
  >
    <li
      class="css-131zfi8"
    >
      Travis Willingham
    </li>
    <li
      class="css-131zfi8"
    >
      Laura Bailey
    </li>
    <li
      class="css-131zfi8"
    >
      Ashleigh Johnson
    </li>
    <li
      class="css-131zfi8"
    >
      Sam Riegal
    </li>
    <li
      class="css-131zfi8"
    >
      Matthew Mercer
    </li>
    <li
      class="css-131zfi8"
    >
      Marisha Ray
    </li>
    <li
      class="css-131zfi8"
    >
      Liam O'Brian
    </li>
    <li
      class="css-131zfi8"
    >
      Talisen Jaffe
    </li>
  </ul>
</div>
`;
